File: app/src/main/java/com/kalenjohnson/chronoduo/saveimport/SnesSrm.java
package com.kalenjohnson.chronoduo.saveimport;

import java.util.ArrayList;
import java.util.List;

/**
 * Parser for SNES Chrono Trigger .srm save files, ported byte-for-byte from
 * tools/saves/snes_srm.py. See REPORT.md #2 for the full slot layout.
 */
public final class SnesSrm {
    public static final int SLOT_SIZE = 0xA00;
    public static final int NUM_SLOTS = 3;
    public static final int CHAR_SIZE = 0x50;
    public static final int NUM_CHARS = 7;
    public static final int CHECKSUM_TABLE_OFFSET = 0x1FF0;
    public static final int LAST_USED_OFFSET = 0x1FE0;

    public static final String[] CHAR_ORDER = {
            "Crono", "Marle", "Lucca", "Robo", "Frog", "Ayla", "Magus",
    };
    public static final String[] NAME_SLOT_ORDER = {
            "Crono", "Marle", "Lucca", "Robo", "Frog", "Ayla", "Magus", "Epoch",
    };

    private SnesSrm() {
    }

    /**
     * 16-bit add-with-carry (65816 ADC loop) over the slot's 1280 LE words,
     * summed from the LAST word (0x9FE) down to the first, final carry-out
     * dropped -- descending order matches 53/53 fixture slots (REPORT.md #2).
     */
    public static int slotChecksum(byte[] slot) {
        int acc = 0;
        int carry = 0;
        for (int i = SLOT_SIZE - 2; i >= 0; i -= 2) {
            int w = (slot[i] & 0xFF) | ((slot[i + 1] & 0xFF) << 8);
            int total = acc + w + carry;
            acc = total & 0xFFFF;
            carry = (total > 0xFFFF) ? 1 : 0;
        }
        return acc;
    }

    /** 0xA0-0xB9 = A-Z, 0xBA-0xD3 = a-z, 0x00 or 0xFF = end/pad. */
    public static String decodeName(byte[] raw, int off, int len) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < len; i++) {
            int b = raw[off + i] & 0xFF;
            if (b == 0 || b == 0xFF) break;
            if (b >= 0xA0 && b <= 0xB9) {
                sb.append((char) ('A' + (b - 0xA0)));
            } else if (b >= 0xBA && b <= 0xD3) {
                sb.append((char) ('a' + (b - 0xBA)));
            } else {
                sb.append('?');
            }
        }
        return sb.toString();
    }

    /** One 0x50-byte character record (REPORT.md #4, SNES column). */
    public static final class SnesChar {
        public final byte[] raw;
        public final int id;
        public final int curHp, maxHp, curMp, maxMp;
        public final int power, stamina, speed, magic, hit, evade, mdef, level;
        public final int exp;
        public final int tpNext;
        public final int helmetId, armorId, weaponId, accessoryId;
        public final int expNext;
        public final int field2d;

        private SnesChar(byte[] raw, int id, int curHp, int maxHp, int curMp, int maxMp,
                          int power, int stamina, int speed, int magic, int hit, int evade, int mdef, int level,
                          int exp, int tpNext, int helmetId, int armorId, int weaponId, int accessoryId,
                          int expNext, int field2d) {
            this.raw = raw;
            this.id = id;
            this.curHp = curHp;
            this.maxHp = maxHp;
            this.curMp = curMp;
            this.maxMp = maxMp;
            this.power = power;
            this.stamina = stamina;
            this.speed = speed;
            this.magic = magic;
            this.hit = hit;
            this.evade = evade;
            this.mdef = mdef;
            this.level = level;
            this.exp = exp;
            this.tpNext = tpNext;
            this.helmetId = helmetId;
            this.armorId = armorId;
            this.weaponId = weaponId;
            this.accessoryId = accessoryId;
            this.expNext = expNext;
            this.field2d = field2d;
        }

        static SnesChar parse(byte[] raw) {
            if (raw.length != CHAR_SIZE) throw new IllegalArgumentException("bad char record size");
            int power = raw[0x0B] & 0xFF;
            int stamina = raw[0x0C] & 0xFF;
            int speed = raw[0x0D] & 0xFF;
            int magic = raw[0x0E] & 0xFF;
            int hit = raw[0x0F] & 0xFF;
            int evade = raw[0x10] & 0xFF;
            int mdef = raw[0x11] & 0xFF;
            int level = raw[0x12] & 0xFF;
            int exp = (raw[0x13] & 0xFF) | ((raw[0x14] & 0xFF) << 8) | ((raw[0x15] & 0xFF) << 16);
            return new SnesChar(
                    raw.clone(),
                    raw[0x00] & 0xFF,
                    u16(raw, 0x03), u16(raw, 0x05), u16(raw, 0x07), u16(raw, 0x09),
                    power, stamina, speed, magic, hit, evade, mdef, level,
                    exp, u16(raw, 0x16),
                    raw[0x27] & 0xFF, raw[0x28] & 0xFF, raw[0x29] & 0xFF, raw[0x2A] & 0xFF,
                    u16(raw, 0x2B), raw[0x2D] & 0xFF
            );
        }
    }

    /** (id, count) inventory entry. */
    public static final class ItemStack {
        public final int id, count;

        public ItemStack(int id, int count) {
            this.id = id;
            this.count = count;
        }
    }

    public static final class SnesSlot {
        public final byte[] raw;
        public final List<ItemStack> items;
        public final SnesChar[] chars; // 7, Crono..Magus
        public final byte[] techBlock; // 45 bytes at 0x430
        public final byte[] party; // 3 bytes
        public final byte[] reserve; // 6 bytes
        public final int recruitedMask;
        public final String[] names; // 8 decoded strings
        public final int gold;
        public final int playHours, playMinutes, playSeconds, playTimeSeconds;
        public final int chapterByte;
        public final int locationId, x, y;
        public final byte[] flags; // 512 bytes at 0x602

        private SnesSlot(byte[] raw, List<ItemStack> items, SnesChar[] chars, byte[] techBlock,
                          byte[] party, byte[] reserve, int recruitedMask, String[] names, int gold,
                          int playHours, int playMinutes, int playSeconds, int playTimeSeconds,
                          int chapterByte, int locationId, int x, int y, byte[] flags) {
            this.raw = raw;
            this.items = items;
            this.chars = chars;
            this.techBlock = techBlock;
            this.party = party;
            this.reserve = reserve;
            this.recruitedMask = recruitedMask;
            this.names = names;
            this.gold = gold;
            this.playHours = playHours;
            this.playMinutes = playMinutes;
            this.playSeconds = playSeconds;
            this.playTimeSeconds = playTimeSeconds;
            this.chapterByte = chapterByte;
            this.locationId = locationId;
            this.x = x;
            this.y = y;
            this.flags = flags;
        }

        static SnesSlot parse(byte[] raw) {
            if (raw.length != SLOT_SIZE) throw new IllegalArgumentException("bad slot size");
            List<ItemStack> items = new ArrayList<>();
            for (int i = 0; i < 256; i++) {
                int id = raw[i] & 0xFF;
                if (id != 0) items.add(new ItemStack(id, raw[0x100 + i] & 0xFF));
            }

            SnesChar[] chars = new SnesChar[NUM_CHARS];
            for (int i = 0; i < NUM_CHARS; i++) {
                byte[] rec = new byte[CHAR_SIZE];
                System.arraycopy(raw, 0x200 + i * CHAR_SIZE, rec, 0, CHAR_SIZE);
                chars[i] = SnesChar.parse(rec);
            }

            byte[] techBlock = new byte[45];
            System.arraycopy(raw, 0x430, techBlock, 0, 45);
            byte[] party = new byte[3];
            System.arraycopy(raw, 0x580, party, 0, 3);
            byte[] reserve = new byte[6];
            System.arraycopy(raw, 0x583, reserve, 0, 6);
            int recruitedMask = raw[0x5AF] & 0xFF;
            String[] names = new String[8];
            for (int i = 0; i < 8; i++) names[i] = decodeName(raw, 0x5B0 + i * 6, 6);
            int gold = (raw[0x5E0] & 0xFF) | ((raw[0x5E1] & 0xFF) << 8) | ((raw[0x5E2] & 0xFF) << 16);

            int sec = raw[0x5E4] & 0xFF;
            int minOnes = raw[0x5E5] & 0xFF;
            int minTens = raw[0x5E6] & 0xFF;
            int hrOnes = raw[0x5E7] & 0xFF;
            int hrTens = raw[0x5E8] & 0xFF;
            int minutes = minTens * 10 + minOnes;
            int hours = hrTens * 10 + hrOnes;
            int playTimeSeconds = hours * 3600 + minutes * 60 + sec;

            int locationId = u16(raw, 0x5F3);
            int x = u16(raw, 0x5F5);
            int y = u16(raw, 0x5F7);
            byte[] flags = new byte[512];
            System.arraycopy(raw, 0x602, flags, 0, 512);

            return new SnesSlot(raw.clone(), items, chars, techBlock, party, reserve, recruitedMask, names, gold,
                    hours, minutes, sec, playTimeSeconds, hrOnes, locationId, x, y, flags);
        }
    }

    /** Return value of {@link #parseSrm}: 3 slots (null = unused), last-used marker, stored checksums. */
    public static final class SrmFile {
        public final SnesSlot[] slots; // length 3, entries may be null
        public final int lastUsed;
        public final int[] checksums; // length 3

        SrmFile(SnesSlot[] slots, int lastUsed, int[] checksums) {
            this.slots = slots;
            this.lastUsed = lastUsed;
            this.checksums = checksums;
        }
    }

    public static SrmFile parseSrm(byte[] data) {
        if (data.length != 8192) {
            throw new IllegalArgumentException("expected 8192-byte .srm, got " + data.length);
        }
        int[] checksums = new int[3];
        for (int i = 0; i < 3; i++) checksums[i] = u16(data, CHECKSUM_TABLE_OFFSET + i * 2);
        int lastUsed = data[LAST_USED_OFFSET] & 0xFF;
        SnesSlot[] slots = new SnesSlot[NUM_SLOTS];
        for (int i = 0; i < NUM_SLOTS; i++) {
            byte[] raw = new byte[SLOT_SIZE];
            System.arraycopy(data, i * SLOT_SIZE, raw, 0, SLOT_SIZE);
            boolean unused = true;
            for (int b = 0; b < 16; b++) {
                if (raw[b] != (byte) 0xFF) {
                    unused = false;
                    break;
                }
            }
            slots[i] = unused ? null : SnesSlot.parse(raw);
        }
        return new SrmFile(slots, lastUsed, checksums);
    }

    private static int u16(byte[] a, int off) {
        return (a[off] & 0xFF) | ((a[off + 1] & 0xFF) << 8);
    }
}

File: tools/saves/snes_srm.py
"""Parser for SNES Chrono Trigger .srm save files.

An .srm is 8192 bytes: three 0xA00-byte slots at 0x0000/0x0A00/0x1400, a
last-used-slot marker at 0x1FE0, and three little-endian u16 slot checksums
at 0x1FF0. An unused slot is entirely 0xFF.

See REPORT.md #2 for the full slot layout this module implements.
"""
import struct
import sys
from dataclasses import dataclass, field

SLOT_SIZE = 0xA00
NUM_SLOTS = 3
CHAR_SIZE = 0x50
NUM_CHARS = 7
CHECKSUM_TABLE_OFFSET = 0x1FF0
LAST_USED_OFFSET = 0x1FE0

CHAR_ORDER = ["Crono", "Marle", "Lucca", "Robo", "Frog", "Ayla", "Magus"]
NAME_SLOT_ORDER = ["Crono", "Marle", "Lucca", "Robo", "Frog", "Ayla", "Magus", "Epoch"]


def slot_checksum(slot: bytes) -> int:
    """16-bit add-with-carry (65816 ADC loop) over the slot's 1280 LE words,
    summed from the LAST word (0x9FE) down to the first, final carry-out
    dropped. Word order matters because the carry propagates; forward order
    matches only 43/53 fixture slots, descending order matches 53/53.
    """
    acc = 0
    carry = 0
    for i in range(SLOT_SIZE - 2, -1, -2):
        w = slot[i] | (slot[i + 1] << 8)
        total = acc + w + carry
        acc = total & 0xFFFF
        carry = 1 if total > 0xFFFF else 0
    return acc


def decode_name(raw: bytes) -> str:
    """0xA0-0xB9 = A-Z, 0xBA-0xD3 = a-z, 0x00 or 0xFF = end/pad."""
    out = []
    for b in raw:
        if b in (0, 0xFF):
            break
        if 0xA0 <= b <= 0xB9:
            out.append(chr(ord('A') + (b - 0xA0)))
        elif 0xBA <= b <= 0xD3:
            out.append(chr(ord('a') + (b - 0xBA)))
        else:
            out.append('?')
    return ''.join(out)


@dataclass
class SnesChar:
    """One 0x50-byte character record (REPORT.md #4, SNES column)."""
    raw: bytes
    id: int
    cur_hp: int
    max_hp: int
    cur_mp: int
    max_mp: int
    power: int
    stamina: int
    speed: int
    magic: int
    hit: int
    evade: int
    mdef: int
    level: int
    exp: int
    tp_next: int
    helmet_id: int
    armor_id: int
    weapon_id: int
    accessory_id: int
    exp_next: int
    field_2d: int

    @staticmethod
    def parse(raw: bytes) -> "SnesChar":
        assert len(raw) == CHAR_SIZE
        power, stamina, speed, magic, hit, evade, mdef, level = raw[0x0B:0x13]
        exp = raw[0x13] | (raw[0x14] << 8) | (raw[0x15] << 16)
        return SnesChar(
            raw=bytes(raw),
            id=raw[0x00],
            cur_hp=struct.unpack_from('<H', raw, 0x03)[0],
            max_hp=struct.unpack_from('<H', raw, 0x05)[0],
            cur_mp=struct.unpack_from('<H', raw, 0x07)[0],
            max_mp=struct.unpack_from('<H', raw, 0x09)[0],
            power=power, stamina=stamina, speed=speed, magic=magic,
            hit=hit, evade=evade, mdef=mdef, level=level,
            exp=exp,
            tp_next=struct.unpack_from('<H', raw, 0x16)[0],
            helmet_id=raw[0x27],
            armor_id=raw[0x28],
            weapon_id=raw[0x29],
            accessory_id=raw[0x2A],
            exp_next=struct.unpack_from('<H', raw, 0x2B)[0],
            field_2d=raw[0x2D],
        )


@dataclass
class SnesSlot:
    raw: bytes
    items: list  # list of (id, count), 256 entries (0 id = empty)
    chars: list  # 7 SnesChar, Crono..Magus
    tech_block: bytes  # 45 bytes at 0x430
    party: bytes  # 3 bytes
    reserve: bytes  # 6 bytes
    recruited_mask: int
    names: list  # 8 decoded strings
    gold: int
    play_hours: int
    play_minutes: int
    play_seconds: int
    play_time_seconds: int
    chapter_byte: int  # 0x5E7 -- see caveat below
    location_id: int
    x: int
    y: int
    flags: bytes  # 512 bytes at 0x602

    @staticmethod
    def parse(raw: bytes) -> "SnesSlot":
        assert len(raw) == SLOT_SIZE
        ids = raw[0x000:0x100]
        counts = raw[0x100:0x200]
        items = [(ids[i], counts[i]) for i in range(256) if ids[i] != 0]

        chars = [
            SnesChar.parse(raw[0x200 + i * CHAR_SIZE: 0x200 + (i + 1) * CHAR_SIZE])
            for i in range(NUM_CHARS)
        ]

        tech_block = raw[0x430:0x430 + 45]
        party = raw[0x580:0x583]
        reserve = raw[0x583:0x589]
        recruited_mask = raw[0x5AF]
        names = [decode_name(raw[0x5B0 + i * 6: 0x5B0 + i * 6 + 6]) for i in range(8)]
        gold = raw[0x5E0] | (raw[0x5E1] << 8) | (raw[0x5E2] << 16)

        # Play time per coordinator correction: 0x5E3..0x5E8 are six DIGIT
        # bytes -- frames, seconds, minutes-ones, minutes-tens, hours-ones,
        # hours-tens -- not a packed frames/seconds/minutes/hours quad as
        # REPORT.md #2 originally described. 0x5E7 is the hours-ones digit,
        # not a "story chapter counter".
        frames = raw[0x5E3]
        sec = raw[0x5E4]
        min_ones = raw[0x5E5]
        min_tens = raw[0x5E6]
        hr_ones = raw[0x5E7]
        hr_tens = raw[0x5E8]
        minutes = min_tens * 10 + min_ones
        hours = hr_tens * 10 + hr_ones
        play_time_seconds = hours * 3600 + minutes * 60 + sec

        location_id = struct.unpack_from('<H', raw, 0x5F3)[0]
        x = struct.unpack_from('<H', raw, 0x5F5)[0]
        y = struct.unpack_from('<H', raw, 0x5F7)[0]
        flags = raw[0x602:0x602 + 512]

        return SnesSlot(
            raw=bytes(raw), items=items, chars=chars, tech_block=bytes(tech_block),
            party=bytes(party), reserve=bytes(reserve), recruited_mask=recruited_mask,
            names=names, gold=gold,
            play_hours=hours, play_minutes=minutes, play_seconds=sec,
            play_time_seconds=play_time_seconds,
            chapter_byte=hr_ones, location_id=location_id, x=x, y=y,
            flags=bytes(flags),
        )


def parse_srm(data: bytes):
    """Return (slots, last_used, checksums). slots[i] is None for unused slots."""
    assert len(data) == 8192
    checksums = struct.unpack_from('<3H', data, CHECKSUM_TABLE_OFFSET)
    last_used = data[LAST_USED_OFFSET]
    slots = []
    for i in range(NUM_SLOTS):
        raw = data[i * SLOT_SIZE:(i + 1) * SLOT_SIZE]
        if raw[:16] == b'\xff' * 16:
            slots.append(None)
        else:
            slots.append(SnesSlot.parse(raw))
    return slots, last_used, checksums


def load_srm(path: str):
    with open(path, 'rb') as f:
        data = f.read()
    return parse_srm(data)


if __name__ == '__main__':
    for path in sys.argv[1:]:
        print(f"=== {path} ===")
        slots, last_used, checksums = load_srm(path)
        print(f"  last used slot marker: {last_used}")
        for i, slot in enumerate(slots):
            if slot is None:
                print(f"  slot {i}: unused")
                continue
            stored = checksums[i]
            computed = slot_checksum(slot.raw)
            ok = "OK" if computed == stored else f"MISMATCH (computed {computed:#06x})"
            party_names = [slot.names[cid] for cid in slot.party if cid < 7]
            print(f"  slot {i}: checksum stored={stored:#06x} {ok}")
            print(f"    party: {', '.join(party_names)}")
            print(f"    gold: {slot.gold}  play time: "
                  f"{slot.play_hours:02d}:{slot.play_minutes:02d}:{slot.play_seconds:02d} "
                  f"({slot.play_time_seconds}s)")
            print(f"    location: {slot.location_id}  x,y: {slot.x},{slot.y}")
            print(f"    recruited mask: {slot.recruited_mask:#04x}")
            for c in slot.chars:
                name = slot.names[c.id] if c.id < 8 else f"id{c.id}"
                print(f"    {name}: lvl {c.level} hp {c.cur_hp}/{c.max_hp} "
                      f"mp {c.cur_mp}/{c.max_mp} exp {c.exp}")
